src/lib/data_schema.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2012-2022  Online-Go.com
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * This is the schema for the data functions (e.g. get(), set()).  It defines
+ * all the possible keys as well as the associated value types.
+ */
+
+import { GroupList, ActiveTournamentList } from "./types";
+import { Announcement } from "src/components/Announcements";
+
+interface CachedSchema {
+    groups: GroupList;
+    active_tournaments: ActiveTournamentList;
+
+    // TODO: examine these endpoints and write interfaces for these.
+    config: ConfigSchema /* /login */;
+    ladders: any /* /me/ladders */;
+    challenge_list: any /* /me/challenges */;
+    blocks: any /* /me/blocks */;
+    friends: any /* ui/friends */;
+    group_invitations: any /* me/groups/invitations */;
+}
+
+export interface ConfigSchema {
+    user: any;
+    cdn: string;
+    cdn_release: string;
+    cdn_host: string;
+    release: string;
+    version: string;
+    paypal_server: string;
+    paypal_this_server: string;
+    paypal_email: string;
+    billing_mor_locations: string[];
+    user_jwt: string;
+    stripe_pk: string;
+    chat_auth: string;
+    superchat_auth: string;
+    notification_auth: string;
+    incident_auth: string;
+    aga_rankings_enabled: boolean;
+}
+
+/**
+ * Prefixes every member of a type.
+ *
+ * Example:
+ *
+ * type fooType = Prefixed<{ bar: string; qux: number }, "foo">;
+ *
+ * is equivalent to
+ *
+ * type fooType = { "foo.bar": string; "foo.qux": number };
+ */
+type Prefixed<T, P extends string> = {
+    [K in keyof T as K extends string ? `${P}.${K}` : never]: T[K];
+};
+
+export interface DataSchema
+    extends Prefixed<CachedSchema, "cached">,
+        Prefixed<ConfigSchema, "config"> {
+    user: any;
+    bid: string;
+    theme: string;
+    debug: boolean;
+
+    cached: Partial<CachedSchema>;
+    config: Partial<ConfigSchema>;
+
+    // TODO: make a types for each of these that list the keys explicitly
```

Comment:
Note: even if this PR is accepted, `data` still has a good deal of work before it would be type safe. Most of the prefixed keys in data have been added in a very permissive way to avoid compiler errors.  But I want to get confirmation that this PR is the right direction before adding more schema interfaces.